File: bindings/java/perf/single/Zlink.BindingBench/src/main/java/systems/zlink/perf/single/PerfSpotReqRep.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.perf.single;

import systems.zlink.Context;
import systems.zlink.DealerSocket;
import systems.zlink.Message;
import systems.zlink.PollEventFlag;
import systems.zlink.RouterSocket;
import systems.zlink.perf.PerfSocketPollSet;
import systems.zlink.perf.PerfStopToken;
import systems.zlink.perf.PerfUtil;
import systems.zlink.service.spot.Spot;
import systems.zlink.service.spot.SpotNode;
import java.time.Duration;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.concurrent.atomic.AtomicReference;

final class PerfSpotReqRep {
    private static final String CHANNEL_NAME = "perf.spot.reqrep";

    private PerfSpotReqRep() {
    }

    static PerfUtil.Result run(PerfUtil.Config config) {
        PerfUtil.Metrics metrics = new PerfUtil.Metrics(config);
        AtomicReference<Throwable> responderFailure = new AtomicReference<>();
        String endpoint = normalizeSpotEndpoint(
            PerfUtil.endpoint(config.transport(), "single-spot-reqrep"),
            config.transport());

        try (Context ctx = PerfUtil.newContext(config);
             SpotNode requesterNode = new SpotNode(ctx);
             DealerSocket requesterDealer = new DealerSocket(ctx);
             Spot requester = requesterNode.createSpot();
             RouterSocket responder = new RouterSocket(ctx)) {
            PerfUtil.applySocketOptions(requesterDealer, config);
            PerfUtil.applySocketOptions(responder, config);
            PerfUtil.configureClientTls(requesterDealer, config.transport());
            PerfUtil.configureServerTls(responder, config.transport());
            requesterNode.attachChannelDealerManual(CHANNEL_NAME, requesterDealer);
            responder.bind(endpoint);
            requesterDealer.connect(endpoint);

            Thread responderThread = new Thread(() -> runResponder(
                responder, config.size(), responderFailure),
                "single-spot-reqrep-responder");
            responderThread.start();

            if (!waitForWarmupReady(requester, config.size(),
                    config.connectReadyTimeoutMs())) {
                throw new IllegalStateException("spot reqrep ready timed out");
            }
            settleAfterReady();

            metrics.startActiveWindow();
            long activeEnd = System.nanoTime()
                + config.durationSeconds() * 1_000_000_000L;
            while (System.nanoTime() < activeEnd) {
                try (Message active = PerfUtil.payload(config.size(),
                         (byte) PerfUtil.PHASE_ACTIVE, System.nanoTime())) {
                    PerfUtil.Header reply = requestReply(requester, active,
                        config.size(), Duration.ofSeconds(2));
                    if (reply.phase() == PerfUtil.PHASE_ACTIVE) {
                        metrics.recordNanos(reply.latencyNanos() / 2L);
                    }
                }
            }
            // PERF_SINGLE_TEST_POLICY § 1.4: signal phase end via wire-level
            // stop token. The responder echoes back the stop token, then
            // observes the stop token in its inbound recv and exits.
            stopResponder(requester, config.size());
            PerfUtil.join(responderThread, "spot reqrep responder",
                Duration.ofSeconds(10));
            if (responderFailure.get() != null) {
                throw new IllegalStateException("spot reqrep responder failed",
                    responderFailure.get());
            }
            ctx.shutdown();
            return metrics.finishSingle(config);
        }
    }

    private static void runResponder(RouterSocket responder,
                                     int expectedSize,
                                     AtomicReference<Throwable> responderFailure) {
        try (PerfSocketPollSet pollSet = PerfSocketPollSet.fromSockets(
                List.of((systems.zlink.Socket) responder),
                PollEventFlag.POLLIN)) {
            // PERF_SINGLE_TEST_POLICY § 1.4: signal-driven wait; exit on
            // wire-level stop token after echoing it.
            while (true) {
                pollSet.poll(-1);
                boolean stop = false;
                while (true) {
                    systems.zlink.Received received = PerfUtil.recvNoWait(responder);
                    if (received == null) {
                        break;
                    }
                    try (received) {
                        boolean isStop = PerfStopToken.isStopTokenMessage(
                            received.firstPart());
                        // Snapshot any non-stop frame as the echoed reply.
                        // For the warmup probe and active payload the requester
                        // expects exactly the bytes it sent.
                        try (Message reply = received.firstPart().move()) {
                            received.reply(reply);
                        }
                        if (isStop) {
                            stop = true;
                            break;
                        }
                    }
                }
                if (stop) {
                    return;
                }
            }
        } catch (Throwable ex) {
            responderFailure.compareAndSet(null, ex);
        }
    }

    private static void stopResponder(Spot requester, int size) {
        // Send the stop token through the request/reply channel; the responder
        // echoes it back and then exits.
        try (Message stop = PerfStopToken.newMessage()) {
            try {
                requester.requestChannel(CHANNEL_NAME)
                    .message(stop)
                    .timeout(Duration.ofSeconds(2))
                    .submitAsync()
                    .get(2, TimeUnit.SECONDS)
                    .forEach(Message::close);
            } catch (InterruptedException ex) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("spot reqrep stop interrupted", ex);
            } catch (ExecutionException ex) {
                throw new IllegalStateException("spot reqrep stop failed",
                    ex.getCause());
            } catch (TimeoutException ex) {
                throw new IllegalStateException("spot reqrep stop timed out", ex);
            }
        }
    }

    private static boolean waitForWarmupReady(Spot requester,
                                              int size,
                                              int timeoutMs) {
        long deadlineNs = System.nanoTime()
            + Duration.ofMillis(Math.max(1, timeoutMs)).toNanos();
        while (System.nanoTime() < deadlineNs) {
            try (Message probe = PerfUtil.payload(size,
                     (byte) PerfUtil.PHASE_WARMUP, System.nanoTime())) {
                PerfUtil.Header reply = requestReply(requester, probe, size,
                    Duration.ofMillis(200));
                if (reply.phase() == PerfUtil.PHASE_WARMUP) {
                    return true;
                }
            } catch (Throwable e) {
                System.err.println("[single-spot-reqrep] warmup probe failed: " + e);
                try {
                    Thread.sleep(10L);
                } catch (InterruptedException ex) {
                    Thread.currentThread().interrupt();
                    throw new IllegalStateException("spot reqrep warmup interrupted",
                        ex);
                }
            }
        }
        return false;
    }

    private static PerfUtil.Header requestReply(Spot requester,
                                                Message payload,
                                                int expectedSize,
                                                Duration timeout) {
        List<Message> replyParts;
        try {
            replyParts = requester.requestChannel(CHANNEL_NAME)
                .message(payload)
                .timeout(timeout)
                .submitAsync()
                .get(timeout.toMillis(), TimeUnit.MILLISECONDS);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("spot reqrep request interrupted", ex);
        } catch (ExecutionException ex) {
            throw new IllegalStateException("spot reqrep request failed", ex.getCause());
        } catch (TimeoutException ex) {
            throw new IllegalStateException("spot reqrep request timed out", ex);
        }
        try {
            if (replyParts.isEmpty()) {
                throw new IllegalStateException("spot reqrep reply was empty");
            }
            PerfUtil.Header header = PerfUtil.decodeHeader(replyParts.get(0), expectedSize);
            if (header == null) {
                throw new IllegalStateException("spot reqrep reply header missing");
            }
            return header;
        } finally {
            replyParts.forEach(Message::close);
        }
    }

    private static void settleAfterReady() {
        int settleMs = PerfUtil.intEnv("PERF_SINGLE_SPOT_READY_SETTLE_MS", 1000);
        if (settleMs <= 0) {
            return;
        }
        try {
            Thread.sleep(settleMs);
        } catch (InterruptedException ex) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("spot reqrep settle interrupted", ex);
        }
    }

    private static String normalizeSpotEndpoint(String endpoint, String transport) {
        if (!"tls".equals(transport) && !"wss".equals(transport)) {
            return endpoint;
        }
        return endpoint.replace("://127.0.0.1:", "://localhost:");
    }
}

File: bindings/dotnet/perf/single/Zlink.BindingBench/src/PerfSpotReqRep.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using Systems.Zlink;
using static PerfRunner;

internal static class PerfSpotReqRep
{
    private const string ChannelName = "perf.spot.reqrep";
    private const uint RunId = 1;
    private const uint WarmupPhase = 0;
    private const uint ActivePhase = 1;

    internal static int Run(string transport, int size)
    {
        if (!IsSupportedTransport(transport))
        {
            PrintUnsupported("SPOT_REQREP", transport, size,
                "spot_reqrep_transport_not_supported");
            return 0;
        }

        int durationSeconds = ResolveSingleDurationSeconds();
        int readyTimeoutMs = ResolveSpotReadyTimeoutMs();
        int latencySampleCap = ResolveSingleLatencyCount("SPOT_REQREP");
        string bindEndpoint = EndpointFor(transport, "single-spot-reqrep");
        string connectEndpoint = NormalizeClientEndpoint(bindEndpoint, transport);

        using var ctx = new Context();
        ApplySingleContextOptions(ctx);
        using var requesterNode = new SpotNode(ctx);
        using var requesterDealer = new DealerSocket(ctx);
        using var requester = requesterNode.CreateSpot();
        using var responder = new RouterSocket(ctx);

        try
        {
            ApplySingleSocketOptions(requesterDealer);
            ApplySingleSocketOptions(responder);
            ConfigureTlsClientIfNeeded(requesterDealer, transport);
            ConfigureTlsServerIfNeeded(responder, transport);
            requesterNode.AttachChannelDealerManual(ChannelName, requesterDealer);
            responder.Bind(bindEndpoint);
            requesterDealer.Connect(connectEndpoint);

            Exception? responderFailure = null;
            // PERF_SINGLE_TEST_POLICY § 1.4: wire-level stop token signals
            // the responder to exit its echo loop. The token is sent as
            // the requester's final request so the requester also drains
            // the matching reply via public poller completion.
            var responderThread = new Thread(() =>
            {
                try
                {
                    while (true)
                    {
                        using Received received = responder.Recv();
                        RoutingId routingId = received.RoutingId
                            ?? throw new InvalidOperationException("missing routing id");
                        ulong requestSeq = received.RequestSeq ?? 0UL;
                        ReadOnlySpan<byte> body = received.FirstPart().AsReadOnlySpan();
                        bool isStop = StopToken.IsStopToken(body);

                        using Message reply = received.FirstPart().Move();
                        responder.Reply(routingId, requestSeq, reply);

                        if (isStop)
                            break;
                    }
                }
                catch (Exception ex)
                {
                    Interlocked.CompareExchange(ref responderFailure, ex, null);
                }
            })
            {
                IsBackground = true,
            };
            responderThread.Start();

            if (!WaitForWarmupReady(requester, size, readyTimeoutMs))
                return 2;

            var samples = new List<double>(Math.Max(0, latencySampleCap));
            long sampleSeen = 0;
            uint rng = 0xA341316Cu;
            long measureCount = 0;
            long activeDeadlineTicks = DeadlineTicksFromSeconds(durationSeconds);
            ulong seq = 1;

            while (Stopwatch.GetTimestamp() < activeDeadlineTicks)
            {
                using var active = Message.FromBytes(
                    CreatePayload(size, ActivePhase, seq++));
                PerfMetricHeader reply = RequestReply(requester, size, active,
                    TimeSpan.FromSeconds(2));
                if (reply.Phase != ActivePhase)
                    continue;

                measureCount++;
                ulong nowNs = EpochNs();
                if (nowNs >= reply.SentTsNs)
                {
                    ReservoirSample(samples, (nowNs - reply.SentTsNs) / 2.0,
                        ref sampleSeen, latencySampleCap, ref rng);
                }
            }

            // Final request carries the wire-level stop token so the
            // responder thread exits its loop after sending the reply.
            using (var stopRequest = Message.FromBytes(StopToken.Bytes))
                _ = RequestReplyAllowStopToken(requester, stopRequest,
                    TimeSpan.FromSeconds(2));

            responderThread.Join();
            if (responderFailure != null)
                throw new InvalidOperationException("spot reqrep responder failed",
                    responderFailure);
            if (measureCount <= 0)
                return 2;

            double throughput = measureCount / (double)Math.Max(1, durationSeconds);
            double fallbackLatencyNs = (Math.Max(1.0, durationSeconds)
                * 1_000_000_000.0) / Math.Max(1.0, measureCount * 2.0);
            var latency = ComputeLatencyStats(samples);
            double latencyNs = latency.mean > 0.0 ? latency.mean : fallbackLatencyNs;
            double latencyP95Ns = latency.p95 > 0.0 ? latency.p95 : latencyNs;
            double latencyP99Ns = latency.p99 > 0.0 ? latency.p99 : latencyP95Ns;
            PrintResult("SPOT_REQREP", transport, size, throughput, latencyNs,
                latencyP95Ns, latencyP99Ns, bandwidthMultiplier: 2.0);
            return 0;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine(
                $"single_spot_reqrep_error:exception:{ex.GetType().Name}:{ex.Message}");
            return 2;
        }
    }

    private static bool WaitForWarmupReady(Spot requester, int size, int timeoutMs)
    {
        long deadlineTicks = DeadlineTicksFromMilliseconds(timeoutMs);
        ulong seq = 1;
        while (Stopwatch.GetTimestamp() < deadlineTicks)
        {
            try
            {
                using var probe = Message.FromBytes(
                    CreatePayload(size, WarmupPhase, seq++));
                PerfMetricHeader reply = RequestReply(requester, size, probe,
                    TimeSpan.FromMilliseconds(200));
                if (reply.Phase == WarmupPhase)
                    return true;
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"[single-spot-reqrep] warmup probe failed: {ex.Message}");
                Thread.Sleep(10);
            }
        }

        return false;
    }

    private static PerfMetricHeader RequestReply(Spot requester, int size,
        Message payload, TimeSpan timeout)
    {
        IReadOnlyList<Message> replyParts = requester.RequestChannel(ChannelName)
            .Message(payload)
            .Timeout(timeout)
            .SubmitAsync()
            .GetAwaiter().GetResult();
        try
        {
            if (replyParts.Count == 0)
                throw new InvalidOperationException("spot reqrep reply was empty");

            ReadOnlySpan<byte> body = replyParts[0].AsReadOnlySpan();
            if (TryDecodeMetricHeader(body, out PerfMetricHeader header)
                && header.MsgSize == (uint)size
                && header.RunId == RunId)
            {
                return header;
            }

            throw new InvalidOperationException("spot reqrep reply header invalid");
        }
        finally
        {
            for (int i = 0; i < replyParts.Count; i++)
                replyParts[i].Dispose();
        }
    }

    private static bool RequestReplyAllowStopToken(Spot requester,
        Message payload, TimeSpan timeout)
    {
        IReadOnlyList<Message> replyParts = requester.RequestChannel(ChannelName)
            .Message(payload)
            .Timeout(timeout)
            .SubmitAsync()
            .GetAwaiter().GetResult();
        try
        {
            return replyParts.Count > 0
                && StopToken.IsStopToken(replyParts[0].AsReadOnlySpan());
        }
        finally
        {
            for (int i = 0; i < replyParts.Count; i++)
                replyParts[i].Dispose();
        }
    }

    private static byte[] CreatePayload(int size, uint phase, ulong seq)
    {
        var payload = new byte[Math.Max(size, PerfMetricHeaderSize)];
        Array.Fill(payload, (byte)'a');
        StampMetricHeader(payload.AsSpan(), RunId, phase, size, seq, EpochNs());
        return payload;
    }

    private static string NormalizeClientEndpoint(string endpoint, string transport)
    {
        if (!transport.Equals("tls", StringComparison.OrdinalIgnoreCase)
            && !transport.Equals("wss", StringComparison.OrdinalIgnoreCase))
        {
            return endpoint;
        }

        return endpoint.Replace("://127.0.0.1:", "://localhost:",
            StringComparison.Ordinal);
    }

    private static bool IsSupportedTransport(string transport)
    {
        return string.Equals(transport, "tcp", StringComparison.OrdinalIgnoreCase)
            || string.Equals(transport, "tls", StringComparison.OrdinalIgnoreCase)
            || string.Equals(transport, "ws", StringComparison.OrdinalIgnoreCase)
            || string.Equals(transport, "wss", StringComparison.OrdinalIgnoreCase);
    }
}
